staging/src/k8s.io/component-base/featuregate/feature_gate.go

```diff
@@ -423,6 +423,37 @@ func (f *featureGate) AddVersioned(features map[Feature]VersionedSpecs) error {
 		return fmt.Errorf("cannot add a feature gate after adding it to the flag set")
 	}
 
+	// Validate new specs are well-formed
+	for name, specs := range features {
+		var lastVersion *version.Version
+		var wasBeta, wasGA bool
+		for i, spec := range specs {
+			if spec.Version == nil {
+				return fmt.Errorf("feature %q did not provide a version", name)
+			}
+			// gates that begin as deprecated must indicate their prior state
+			if i == 0 && spec.PreRelease == Deprecated && spec.Version.Minor() != 0 {
+				return fmt.Errorf("feature %q introduced as deprecated must provide a 1.0 entry indicating initial state", name)
+			}
+			if i > 0 {
+				// versions must strictly increase
+				if !lastVersion.LessThan(spec.Version) {
+					return fmt.Errorf("feature %q lists version transitions in non-increasing order (%s >= %s)", name, spec.Version, lastVersion)
+				}
+				// stability must not regress from ga --> {beta,alpha} or beta --> alpha
+				switch {
+				case spec.PreRelease == Alpha && (wasBeta || wasGA):
+					return fmt.Errorf("feature %q regresses stability from more stable level to %s in %s", name, spec.PreRelease, spec.Version)
+				case spec.PreRelease == Beta && wasGA:
+					return fmt.Errorf("feature %q regresses stability from more stable level to %s in %s", name, spec.PreRelease, spec.Version)
+				}
```

Comment:
Nice core set of rules.

Should we also enforce `LockedToDefault: true` for GA here or are rules like that better put into linters?

Reply:
I was also trying to think if there are any hard rules for `PreRelease == Deprecated` transitions we can put in place. It's tricky because I can imagine a situation where we want to deprecate something and then realize there is some reason we can't and so we back away from the deprecation..

Reply:
probably not here... there are a couple weird cases where things hit GA and didn't lock for some reason... trying to keep this very focused on malformed things or things that don't give the system enough information to work properly